### PR TITLE
add a diffusion module

### PR DESCRIPTION
**Pull request by aurelienpierre**
This implements an heat transfer partial derivative equations solver, adapted from https://www.researchgate.net/publication/220663968 and originally meant for image inpainting and reconstruction. However, I have found that it doesn't really reconstruct anything unless you apply a large amount of iterations, which is not compatible with dt's "realtime"-ish purpose. 

The algorithm is adapted for a wavelets framework, which makes it super slow (OpenCL incoming ASAP) but allows to include spatiality in the parameters. It's very close in logic to the filmic highlights reconstruction. Basically, it allows to make pixels spill (or un-spill) on their neighbours.

Screenshot_20210104_014715

The module comes with 4 presets to showcase what it can do, since it's quite close to physics and can be used in a variety of situations:
1. blind deconvolution by gradient descent, following the steepest gradient direction,
2. anisotropic diffusion, or directional blur if you prefer, propagating colors in the perpendicular direction of edges, (so, either
4. joint sharpening and denoising,
5. highlights inpainting (although I'm not convinced by the results).

## Use case 1

The motivation for this module was to spill image colors outside of its borders, as a special effect for prints:

Demi-marathon Hadrien-0044-_DSC0487_04

This is achieved by underlaying a framing module, for which I had to remove the iop_order fences in rawoverexposed.c and finalscale.c. This seems to make dt crash sometimes, even though it's hard to know what is responsible for the crashes these days.

## Use case 2

Deblur. For this, the module is put right after exposure. Before/After 1/After 2/After 3 (images at 1:1):
Demi-marathon Hadrien-0044-_DSC0487_06
Demi-marathon Hadrien-0044-_DSC0487_05
Demi-marathon Hadrien-0044-_DSC0487_07
Demi-marathon Hadrien-0044-_DSC0487_08

Notice that it respects the depth of field fairly well and does not display halos. The filter is applied on linear RGB so it's really suited to remove non-uniform lens blur.

## Caveats

The module is slow… Not much I can do besides providing an OpenCL kernel. 1 to 2 iterations are ok, up to 4 iterations is meh, more than 4 iterations is something you can do before taking a walk.

## How to use

Read the tooltips and use the presets. Details below.

### General

Every order of derivatives comes with 3 sliders:
* strength: 
  * positive values make that derivative smoother (diffusion), 
  * negative values make that derivative more oscillatory (sharpening),
* edges directivity: 
  * high values make the 2D derivatives more directional (anisotropic), following a direction perpendicular to the steepest gradient
  * low values makes the 2D derivatives more uniform (isotropic), in a radial direction, more like usual blurs,
* edges regularization:
  * high values apply an higher penalty to derivatives of high magnitude, to avoid overshoot and gradient reversals around edges,
  * low values apply a lower (or even close to none) penalty to derivatives of high magnitude, allowing stronger effects but also overshoots.
  * low magnitude derivatives (zero and close) are never affected by the regularization parameter. The value of the regularization drives the speed of the transition between the low and high magnitude.

Overshoots cause artifacts, such asnoise, gradient reversals (halos, fringes), etc. If they happen, you need to dampen your derivatives (strength and/or regularization).

The speed of diffusion is a general factor applied to all strengths settings at once. As the number of iterations increases, it's a good idea to decrease the speed to avoid issues.

### First order

First order derivatives are linked to simple diffusion, such as hazing and light leaks. Used in sharpening mode, it's very prone to gradients reversals, so don't force on the strength.

### Second order

The second order is isotropic because it is simply the detail layer of the wavelets separation, that uses a B-spline blur. This setting is exactly similar to the luma curve in contrast equalizer, but the wavelets decomposition is applied in linear RGB instead of Lab. Second order derivatives are linked to sharp edges and acutance (local filter), but also to noise.

### Third order

The third order is linked to the smoothness of the second, so it can be used to alleviate noise when the second order is used in sharpening mode.

### Non-local fourth order

The non-local fourth order will increase or decrease the curvature of the second order by sampling points further away. This can be used to inpaint the texture in damaged areas, or smoothen or increase oscillations. Used with large radii (> 128 px), and small strength, it behaves in the same spirit as the local laplacian for local contrast enhancement.

### Spirit

Start with setting the first and second orders roughly, as they yield the more dramatic results, zoomed-out. When you are sort-of pleased with the result, zoom-in and fine-tune with the third order by watching out for noise. Finish with fourth order for non-local effects like increasing the volume of features.

Then finish with the diffusion speed, to boost or temper the correction. For large deblurrings, inpaintings or joint sharpening + denoising, more iterations will be needed with smaller diffusion speed to avoid overshoots. In the literature, such applications need more than 50 iterations.

By combining the weights of each order, you are virtually able to add or remove any kind of diffusion (atmospheric hazing, frosten glass, lens blur, coma, etc.) or even perform edge-aware surface blurs.

While the same method could be used to denoise, it doesn't take advantage of the inter-channels correlations, so it's no better than the usual denoising modules (but could be modified in Y0U0V0 for that purpose).

### CODE REVIEW

**Comment by aurelienpierre** [comment]
This is ready for review and OpenCL has just been added. For testing, please use the collection of presets provided.

**Comment by johnny-bit** [comment]
Bl;ergh, the CI yells about:

```
In file included from lib/darktable/plugins/introspection_diffuse.c:96:0:
/home/runner/work/darktable/darktable/src/src/iop/diffuse.c:996:15: error: unknown type name ‘cl_int’
 static inline cl_int wavelets_process_cl(const int devid, cl_mem in, cl_mem reconstructed, cl_mem mask,
               ^~~~~~
compilation terminated due to -Wfatal-errors.
```

File: data/kernels/diffuse.cl [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+  {
+    case(DT_ISOTROPY_ISOTROPE):
+    default:
+    {
+      isotrope_laplacian(kern);
+      break;
+    }
+    case(DT_ISOTROPY_ISOPHOTE):
+    {
+      float4 a[2][2] = { { 0.f } };
```
Comment (elstoc) [inline data/kernels/diffuse.cl:193]:
```suggestion
      float4 a[2][2] = { { (float4)0.f } };
```
As discussed. Similarly on line 200

**Comment by matze** [comment]
I suppose there is some other wrong initialization besides those vectors. When zoomed in once (but only on this zoom level), I sometimes get this (AMD but fast math is not the issue):

diffusion

Otherwise, great improvements! With GPU acceleration it is really nice to use and results are outstanding as usual.

File: data/kernels/diffuse.cl (hunk 1 above)
Comment (aurelienpierre) [reply data/kernels/diffuse.cl]: fixed

**Comment by aurelienpierre** [comment]
@matze do you have a more specific reproducer ?

**Comment by elstoc** [comment]
@aurelienpierre: Please can you update the documentation in the first post as some of your sliders now have different names. Also, some of the newer sliders don't have tooltips:

- 2nd order (laplacian)
- edge sensitivity
- edge threshold

**Comment by aurelienpierre** [comment]
Doc is here : https://github.com/darktable-org/dtdocs/pull/290

**Comment by ralfbrown** [comment]
There seems to be a small problem with the luminance masking threshold slider.  The value is displayed to two decimal places, but it seems to be rounded to the nearest full percent. 800% seems a bit high as the default soft bound; would 200% be better?  That still allows +1EV adjustment prior to the module before one needs to exceed the soft limit.

In addition, (CPU codepath) when nonzero, all pixels above the threshold look like they've been posterized to one or two bits per channel.

**Comment by aurelienpierre** [comment]
> 800% seems a bit high as the default soft bound; would 200% be better? That still allows +1EV adjustment prior to the module before one needs to exceed the soft limit.

That depends on the exposure bias you used… That's cosmetic anyway, we can always change that later.

> In addition, (CPU codepath) when nonzero, all pixels above the threshold look like they've been posterized to one or two bits per channel.

When using non zero threshold, the masked parts are inited with noise to start inpainting. Is that what you refer to ?

**Comment by ralfbrown** [comment]
Denoise preset but with masking threshold changed to 10%.  A bit more visible using the "simulate line drawing" preset and a less using the "watercolor" and "surface blur" presets.
Diffuse-LuminanceMask-10percent

**Comment by matze** [comment]
> @matze do you have a more specific reproducer ?

No. I updated your branch and rebased it on master and it's gone. But I also see the color "noise" described by @ralfbrown.

**Comment by aurelienpierre** [comment]
The noise is normal, that's how you initialize inpainting by PDE, which the use case for masking highlights like this.

**Comment by elstoc** [comment]
I'm testing a single image with the "lens deblur: hard" preset. If I double-click to toggle lighttable to to darkroom and back, a few times, eventually it quits with a segfault. Sometimes it happens the first time I go into the darkroom and sometimes it takes a few tries. But always segfaults in the end. OpenCL enabled using a 2Gb nVidia GeForce GTX 770 with legacy (340.108) drivers.

File: src/iop/diffuse.c [hunk 2]
```diff
@@ -211,16 +211,34 @@ void init_presets(dt_iop_module_so_t *self)
...
+  p.anisotropy_first = 0.f;
+  p.anisotropy_second = 0.f;
+  p.anisotropy_third = 0.f;
+  p.anisotropy_fourth = +5.f;
+
+  p.first = -0.02f;
+  p.second = 0.0f;
+  p.third = 0.0 f;
+  p.fourth = +0.05f;
+  dt_gui_presets_add_generic(_("denoise: soft"), self->op, self->version(), &p, sizeof(p), 1, DEVELOP_BLEND_CS_RGB_SCENE);
```
Comment (ralfbrown) [inline src/iop/diffuse.c:241]: same name as the previous preset; presumably you forgot to edit one of them.

File: src/iop/diffuse.c [hunk 3]
```diff
@@ -1016,12 +1052,18 @@ static inline cl_int wavelets_process_cl(const int devid, cl_mem in, cl_mem reco
           compute_anisotropy_factor(data->anisotropy_third),
           compute_anisotropy_factor(data->anisotropy_fourth) };
 
+  fprintf(stdout, "anisotropy : %f ; %f ; %f ; %f \n",
```
Comment (ralfbrown) [inline src/iop/diffuse.c:1055]: leftover debug statement?

**Comment by elstoc** [comment]
As discussed I've retested in master and I get the same segfault so I don't think it can be this PR

File: src/iop/diffuse.c [hunk 4]
```diff
@@ -0,0 +1,1476 @@
...
+  p.regularization = 1.f;
+
+  p.anisotropy_first = 0.f;
+  p.anisotropy_second = 0.f;
+  p.anisotropy_third = 0.f;
+  p.anisotropy_fourth = +5.f;
+
+  p.first = -0.02f;
+  p.second = 0.0f;
+  p.third = 0.0 f;
```
Comment (elstoc) [inline src/iop/diffuse.c:239]:
```suggestion
  p.third = 0.0f;
```

File: src/iop/diffuse.c [hunk 5]
```diff
@@ -0,0 +1,1481 @@
...
+
+
+#ifdef _OPENMP
+#pragma omp declare simd
+#endif
+static inline dt_isotropy_t check_isotropy_mode(const float anisotropy)
+{
+  // user param is negative, positive or zero. The sign encodes the direction of diffusion, the magnitude encodes the ratio of anisotropy
+  // ultimately, the anisotropy factor needs to be positive before going into the exponential
+  if(anisotropy == 0.f) return DT_ISOTROPY_ISOTROPE;
```
Comment (TurboGit) [inline src/iop/diffuse.c:109]: style: better having all return on a new line to have them aligned and more readable.

File: src/iop/diffuse.c [hunk 6]
```diff
@@ -0,0 +1,1481 @@
...
+  p.anisotropy_second = +0.f;
+  p.anisotropy_third = +2.f;
+  p.anisotropy_fourth = +2.f;
+
+  p.first = +0.0f;
+  p.second = +0.0f;
+  p.third = +0.02f;
+  p.fourth = +0.02f;
+  dt_gui_presets_add_generic(_("inpaint highlights"), self->op, self->version(), &p, sizeof(p), 1, DEVELOP_BLEND_CS_RGB_SCENE);
+
```
Comment (TurboGit) [inline src/iop/diffuse.c:377]: 2 superfluous empty lines

File: src/iop/diffuse.c [hunk 7]
```diff
@@ -0,0 +1,1481 @@
...
+          const size_t k_index = (row * width + col) * 4;
+
+          const float DT_ALIGNED_ARRAY filter[FSIZE]
+              = { 1.0f / 16.0f, 4.0f / 16.0f, 6.0f / 16.0f, 4.0f / 16.0f, 1.0f / 16.0f };
+          const float filters = filter[ii] * filter[jj];
+
+          for_four_channels(c, aligned(in : 64) aligned(acc : 16)) acc[c] += filters * in[k_index + c];
+        }
+
+      for_four_channels(c, aligned(in, HF, LF : 64) aligned(acc : 16))
```
Comment (TurboGit) [inline src/iop/diffuse.c:453]: I kind of remember that for_four_channels() was not OK when inside an already SIMD loop, no?

File: src/iop/diffuse.c [hunk 8]
```diff
@@ -0,0 +1,1481 @@
...
+
+        for(size_t ii = 0; ii < 3; ii++)
+          for(size_t jj = 0; jj < 3; jj++)
+            for_four_channels(c, aligned(neighbour_pixel_LF, neighbour_pixel_HF, HF, LF : 64))
+            {
+              neighbour_pixel_HF[3 * ii + jj][c] = HF[(i_neighbours[ii] * width + j_neighbours[jj]) * 4 + c];
+              neighbour_pixel_LF[3 * ii + jj][c] = LF[(i_neighbours[ii] * width + j_neighbours[jj]) * 4 + c];
+            }
+
+        for_four_channels(c, aligned(neighbour_pixel_LF, neighbour_pixel_HF, out, LF, HF : 64) \
```
Comment (TurboGit) [inline src/iop/diffuse.c:676]: same comment as above. but maybe my recollection is wrong, at least this compiles on my side.

File: src/iop/diffuse.c [hunk 9]
```diff
@@ -0,0 +1,1481 @@
...
+          out[index + c] = HF[index + c] + LF[index + c];
+      }
+    }
+}
+
+static inline float compute_anisotropy_factor(const float user_param)
+{
+  // compute the K param in c evaluation from https://www.researchgate.net/publication/220663968
+  // but in a perceptually-even way, for better GUI interaction
+  if(user_param == 0.f) return FLT_MAX;
```
Comment (TurboGit) [inline src/iop/diffuse.c:753]: style: return on new line

**Review by TurboGit — CHANGES_REQUESTED** [review]
Some minor style changes. A question about nested SIMD.

And we need to add the new file diffuse.c into po/POTFILES.in.

We will also need a regression test. I can do that later if you want. Thanks.

File: src/iop/diffuse.c [hunk 10]
```diff
@@ -0,0 +1,1481 @@
+/*
+   This file is part of darktable,
+   Copyright (C) 2019-2020 darktable developers.
```
[the remark this answers is not in the document]
Comment (aurelienpierre) [reply src/iop/diffuse.c:3]: done, copy-pasting error

File: src/iop/diffuse.c (hunk 5 above)
Comment (aurelienpierre) [reply src/iop/diffuse.c:109]: done

File: src/iop/diffuse.c (hunk 6 above)
Comment (aurelienpierre) [reply src/iop/diffuse.c:377]: done

File: src/iop/diffuse.c (hunk 7 above)
Comment (aurelienpierre) [reply src/iop/diffuse.c:453]: The parent loop is not SIMD, no.

File: src/iop/diffuse.c (hunk 8 above)
Comment (aurelienpierre) [reply src/iop/diffuse.c:676]: same, if you look carefully, we are not within a SIMD loop

File: src/iop/diffuse.c (hunk 9 above)
Comment (aurelienpierre) [reply src/iop/diffuse.c:753]: done

**Comment by aurelienpierre** [comment]
Changes made and addition to POTFILES done (I wasn't aware of this one).

I will do the regression test too, because it needs to be aggressive enough without being to slow to run.

**Comment by aurelienpierre** [comment]
See https://github.com/darktable-org/darktable-tests/pull/7

File: src/iop/diffuse.c (hunk 7 above)
Comment (TurboGit) [reply src/iop/diffuse.c:453]: Oh, I've been mistaken by the fact that you have a `schedule(simd: static)` directive and now I'm wondering if this has some semantics if the loop is not simd? Can you clarify? This is out of my knowledge :)

Comment (aurelienpierre) [reply src/iop/diffuse.c:453]: SIMD loops have an explicit `for parallel simd` statement. The `schedule()` statement is only defining how the parallelization should split the loop between threads. `static:simd` means the chunks have the same size as the native SIMD vector, to allow in-thread vectorization when possible.

**Review by TurboGit — APPROVED** [review]
Sounds good to me, Thanks !

**Comment by TurboGit** [comment]
@aurelienpierre : Just noticed that the module has not been added into any module group and so not directly visible. Can you add it ? TIA.

**Comment by jenshannoschwalm** [comment]
Haven't tested this before --> it's huge!

**Comment by aurelienpierre** [comment]
@TurboGit is it not enough to declare `int default_group()` in module ?

**Comment by TurboGit** [comment]
Sorry was not clear. I was talking about the presets.

**Comment by s7habo** [comment]
Excellent module! You have managed to create another masterpiece! 

One thought though - would it make sense to include the color desaturation option? 
With the help of very rich diffusion function combined with blend modes, you get even much nicer results than with lowpass module. Sharpening also has great potential in this regard.

**Comment by aurelienpierre** [comment]
@s7habo Thanks, but this module does physics and knows no color, so a definition of saturation that is not entirely bullshit is not possible. Also, chromaticity is always low-frequency, so it's not really the scope of what we do here.
